// @tago-io/sdk/out/modules/Migration/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
function convert(oldWidget) {
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        analysis_run: oldWidget?.analysis_run,
        dashboard: oldWidget.dashboard,
        data: oldWidget?.data || [],
        display: {
            header_buttons: oldDisplay.header_buttons || [],
            header_size: oldDisplay?.header_size,
            help: oldDisplay.help || "",
            parameters: oldDisplay?.parameters || [],
            theme: {
                color: {
                    background: null,
                },
            },
            url: oldDisplay?.url || "",
            user: oldDisplay?.user,
            variables: oldDisplay?.variables || [],
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "iframe",
    };
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    const isOld = !!(widget?.display?.watermark !== undefined);
    return isOld;
}
exports.isOldStructure = isOldStructure;
