// @tago-io/sdk/out/modules/Migration/tile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
const common_1 = require("./common");
function convert(oldWidget) {
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        dashboard: oldWidget.dashboard,
        data: [],
        display: {
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            image_type: oldDisplay?.icon_color ? "icon" : "image",
            image_url: oldDisplay?.icon_url || "",
            occupy_whole_widget: oldDisplay?.fit_image,
            opacity: oldDisplay?.opacity,
            theme: {
                color: {
                    background: oldDisplay?.background_color || null,
                    header: null,
                    hover: null,
                    icon: oldDisplay?.icon_color || null,
                    title: oldDisplay?.label_color || null,
                },
            },
            title: oldDisplay?.label_button || "",
            url: (0, common_1.removeHttpFromURL)(oldDisplay?.link) || "",
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "tile",
    };
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    const isOld = !!(widget?.display?.background_color ||
        widget?.display?.fit_image ||
        widget?.display?.label_color ||
        widget?.display?.label_button ||
        widget?.display?.link);
    return isOld;
}
exports.isOldStructure = isOldStructure;
