// @tago-io/sdk/out/modules/Migration/video.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
function convertSource(oldDisplay) {
    return {
        static_video: oldDisplay?.static_media || oldDisplay?.static_image,
        type: "static",
    };
}
function convert(oldWidget) {
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        dashboard: oldWidget.dashboard,
        display: {
            auto_play: true,
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            initially_muted: true,
            loop: true,
            object_fit: "contain",
            occupy_whole_widget: false,
            show_controls: true,
            source: convertSource(oldDisplay),
            theme: {
                color: {
                    background: null,
                    header: null,
                },
            },
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "video",
    };
    if (Array.isArray(oldWidget.data)) {
        newStructure.data = oldWidget.data;
    }
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    const isOld = !!(widget?.display?.type_media ||
        widget?.display?.type_image ||
        widget?.display?.static_media ||
        widget?.display?.static_image ||
        widget?.display?.conditions);
    return isOld;
}
exports.isOldStructure = isOldStructure;
