// @tago-io/sdk/out/modules/Services/PDF.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class PDFService extends TagoIOModule_1.default {
    /**
     * Generate a PDF from html, url or base64
     *
     */
    async generate(params) {
        try {
            const result = await (0, axios_1.default)({
                method: "POST",
                url: "https://pdf.middleware.tago.io",
                data: params,
                headers: {
                    token: this.params.token,
                },
            });
            return result?.data;
        }
        catch (error) {
            return Promise.reject(error?.response?.data || { result: false, stuats: "unknown error" });
        }
    }
}
exports.default = PDFService;
